src/common/variant.h

```diff
@@ -148,16 +161,26 @@ class variant final
 
     /// apply a visitor to the variant
     template <typename VisitorT>
-    typename VisitorT::result_type visit(VisitorT &&visitor)
+    decltype(auto) visit(VisitorT &&visitor) // decltype(auto) since it forwards the return ref type correctly
     {
         return boost::apply_visitor(std::forward<VisitorT>(visitor), m_value);
     }
     template <typename VisitorT>
-    typename VisitorT::result_type visit(VisitorT &&visitor) const
+    decltype(auto) visit(VisitorT &&visitor) const // decltype(auto) since it forwards the return ref type correctly
     {
         return boost::apply_visitor(std::forward<VisitorT>(visitor), m_value);
     }
 
+    /// value initialize the variant based on a type index
+    void value_initialize_to_type_index(const int which)
```

Comment:
In what situations would you need to construct based on index rather than type?

Reply:
deserialization

Reply:
Something similar is already done in the current variant de serialization code:  https://github.com/monero-project/monero/blob/059028a30a8ae9752338a7897329fe8012a310d5/src/serialization/variant.h#L74-L80